finality-aleph/src/finalization.rs

```diff
@@ -77,194 +80,133 @@ where
     }
 }
 
-/// Given hash `last_finalized` and `AlephDataFor` `new_data` of two blocks, returns
-/// Some(new_data) if the block hash represented by new_data is a descendant of last_finalized
-/// (and the new_data.number is correct). Otherwise it outputs None.
-pub(crate) fn should_finalize<B, C>(
-    last_finalized: B::Hash,
-    new_data: AlephDataFor<B>,
+/// Checks if the data is available and that the block numbers match with the ones provided.
+pub(crate) fn is_proposal_available_and_correct<B, C>(
+    proposal: &AlephProposal<B>,
     client: &C,
-    last_block_in_session: NumberFor<B>,
-) -> Option<AlephDataFor<B>>
+    session_boundaries: SessionBoundaries<B>,
+) -> bool
 where
     B: Block,
     C: HeaderBackend<B>,
 {
-    // this early return is for optimization reasons only.
-    if new_data.hash == last_finalized {
-        return None;
+    let first_block_in_session = session_boundaries.first_block;
+    let last_block_in_session = session_boundaries.last_block;
+    if proposal.branch.is_empty() {
+        error!(target:"afa", "Incorrect proposal {:?} passed through data availability", proposal);
+        return false;
     }
-
-    if new_data.number > last_block_in_session {
-        return None;
+    if proposal.branch.len() > MAX_DATA_BRANCH_LEN as usize {
+        error!(target:"afa", "Incorrect proposal {:?} passed through data availability", proposal);
+        return false;
     }
 
-    let last_finalized_number = match client.number(last_finalized) {
-        Ok(Some(number)) => number,
-        _ => {
-            error!(target: "afa", "No block number for {}", last_finalized);
-            return None;
-        }
-    };
-
-    if let Ok(Some(header)) = client.header(BlockId::Hash(new_data.hash)) {
-        if *header.number() != new_data.number {
-            warn!(target: "afa", "Incorrect number for hash {}. Got {}, should be {}", new_data.hash, new_data.number, header.number());
-            return None;
-        }
-    } else {
-        warn!(target: "afa", "No header for hash {}", new_data.hash);
-        return None;
+    // Since `proposal.branch.len() == 0` is ruled out above, so we can safely do `proposal.branch.len() - 1`
+    let branch_base_num =
+        proposal.number - <NumberFor<B>>::saturated_from(proposal.branch.len() - 1);
+    if branch_base_num < first_block_in_session {
+        warn!(target:"afa", "Incorrect proposal {:?}. Too low num {:?} first_block {:?}", proposal, proposal.number, first_block_in_session);
+        return false;
+    }
+    if proposal.number > last_block_in_session {
+        warn!(target:"afa", "Incorrect proposal {:?}. Too high num {:?} last_block {:?}", proposal, proposal.number, last_block_in_session);
+        return false;
     }
```

Comment:
matter of taste, but I would prefer adding respective method to `SessionBoundaries`

Reply:
I added a method to `AlephProposal` instead. But should be fine.